File: git-epic-creator/services/code_graph_ingestion_service/src/core/repo_materializer.py
"""Deterministic repo materialization (Task 03).

- ZIP: safe, deterministic extraction (no zip-slip)
- Git: clone + optional ref checkout (dulwich; deterministic workspace path)
"""

from __future__ import annotations

import io
import os
import posixpath
import shutil
import stat
import time
import zipfile
from dataclasses import dataclass
from pathlib import Path

import safezipfile
from dulwich import porcelain
from dulwich.repo import Repo


class ZipSlipError(ValueError):
    """Raised when a ZIP entry would escape the destination directory."""


class UnsafeZipEntryError(ValueError):
    """Raised when a ZIP entry is unsafe (e.g., symlink)."""


def _is_symlink(info: zipfile.ZipInfo) -> bool:
    # External attributes are platform-specific; this is best-effort and safe.
    # When created on Unix, the high 16 bits encode st_mode.
    mode = (info.external_attr >> 16) & 0xFFFF
    return stat.S_IFMT(mode) == stat.S_IFLNK


def _normalized_zip_relpath(name: str) -> str:
    """Normalize a ZIP entry name into a safe POSIX relative path."""

    # ZIP spec uses forward slashes, but Windows tools may emit backslashes.
    p = name.replace("\\", "/")
    p = posixpath.normpath(p)

    # Drop leading "./"
    if p.startswith("./"):
        p = p[2:]

    # Reject absolute paths and drive-letter-ish paths.
    if p.startswith("/") or p.startswith("\\") or ":" in p.split("/", 1)[0]:
        raise ZipSlipError(f"Unsafe absolute path in zip entry: {name!r}")

    parts = [part for part in p.split("/") if part not in ("", ".")]
    if any(part == ".." for part in parts):
        raise ZipSlipError(f"Unsafe parent traversal in zip entry: {name!r}")

    rel = "/".join(parts)
    if rel in ("", "."):
        return ""
    return rel


def materialize_zip_bytes(*, zip_bytes: bytes, dest_dir: Path) -> Path:
    """Extract a ZIP payload deterministically into `dest_dir`."""

    # Deterministic: ensure the output directory is empty.
    if dest_dir.exists():
        _rmtree_retry(dest_dir)
    dest_dir.mkdir(parents=True, exist_ok=True)

    # Use a hardened ZipFile implementation to mitigate zip-slip/zip-bomb risks.
    # We still enforce deterministic ordering + explicit symlink rejection.
    with safezipfile.ZipFile(io.BytesIO(zip_bytes)) as zf:
        infos = list(zf.infolist())

        # Deterministic extraction order: sort by normalized repo-relative path.
        def sort_key(info: zipfile.ZipInfo) -> str:
            try:
                return _normalized_zip_relpath(info.filename)
            except Exception:
                # Still stable: fall back to raw filename so failures are deterministic.
                return info.filename.replace("\\", "/")

        ordered: list[zipfile.ZipInfo] = []
        for info in sorted(infos, key=sort_key):
            rel = _normalized_zip_relpath(info.filename)
            if rel == "":
                continue

            if _is_symlink(info):
                raise UnsafeZipEntryError(f"Symlinks are not allowed in zip: {info.filename!r}")
            ordered.append(info)

        # Note: safezipfile adds safety limits (max_files/max_total_size/etc) in extractall().
        # Passing `members=` preserves our deterministic extraction order.
        zf.extractall(path=dest_dir, members=ordered)

    return dest_dir


@dataclass(frozen=True)
class GitMaterializationResult:
    repo_root: Path
    head_commit: str | None


def materialize_git(
    *,
    git_url: str,
    ref: str | None,
    dest_dir: Path,
) -> GitMaterializationResult:
    """Clone a git repo into `dest_dir` and optionally checkout `ref`."""

    # Deterministic: ensure the output directory is empty.
    if dest_dir.exists():
        _rmtree_retry(dest_dir)
    dest_dir.parent.mkdir(parents=True, exist_ok=True)

    # Keep clone deterministic and quiet.
    err = io.BytesIO()

    porcelain.clone(git_url, target=str(dest_dir), checkout=True, errstream=err)

    if ref:
        # Accept branch name, tag, or commit-ish. `force=True` keeps re-runs stable.
        porcelain.checkout(str(dest_dir), ref, force=True)

    head_commit: str | None = None
    try:
        repo = Repo(str(dest_dir))
        head = repo.head()  # bytes (hex sha) when present
        if isinstance(head, (bytes, bytearray)):
            head_commit = head.decode("ascii", errors="replace")
        else:
            head_commit = str(head)
    except Exception:
        # Materialization succeeded; HEAD might be absent for unusual repos.
        head_commit = None

    return GitMaterializationResult(repo_root=dest_dir, head_commit=head_commit)


def _rmtree_retry(path: Path, *, attempts: int = 30, delay_sec: float = 0.1) -> None:
    """Robust rmtree for Windows file-lock flakiness."""

    last_err: Exception | None = None

    def _onerror(func, p, exc_info):
        # Try to make the path writable and retry once.
        os.chmod(p, stat.S_IWRITE)
        func(p)

    for _ in range(attempts):
        try:
            shutil.rmtree(path, onerror=_onerror)
            return
        except Exception as e:
            last_err = e
            time.sleep(delay_sec)
    if last_err:
        raise last_err

File: git-epic-creator/services/code_graph_ingestion_service/src/core/repo_materializer.rs
use git2::{Repository, build::CheckoutBuilder};
use std::io::Cursor;
use std::path::{Path, PathBuf};
use std::thread::sleep;
use std::time::Duration;

use crate::core::zip_utils::{ZipPathError, normalized_zip_relpath};

#[derive(Debug, thiserror::Error)]
pub enum MaterializeError {
    #[error(transparent)]
    ZipPath(#[from] ZipPathError),
    #[error("unsafe zip entry (symlink): {0}")]
    UnsafeZipEntry(String),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Zip(#[from] zip::result::ZipError),
    #[error(transparent)]
    Git(#[from] git2::Error),
}

fn rmtree_retry(path: &Path, attempts: usize, delay: Duration) -> std::io::Result<()> {
    let mut last_err: Option<std::io::Error> = None;
    for _ in 0..attempts {
        match std::fs::remove_dir_all(path) {
            Ok(()) => return Ok(()),
            Err(e) => {
                last_err = Some(e);
                sleep(delay);
            }
        }
    }
    Err(last_err.unwrap_or_else(|| std::io::Error::other("failed to remove dir")))
}

pub fn materialize_zip_bytes(
    zip_bytes: &[u8],
    dest_dir: &Path,
) -> Result<PathBuf, MaterializeError> {
    // Deterministic: ensure the output directory is empty.
    if dest_dir.exists() {
        rmtree_retry(dest_dir, 30, Duration::from_millis(100))?;
    }
    std::fs::create_dir_all(dest_dir)?;

    let mut archive = zip::ZipArchive::new(Cursor::new(zip_bytes))?;

    // Deterministic extraction order: sort by normalized repo-relative path.
    let mut ordered: Vec<(String, usize)> = Vec::new();
    for i in 0..archive.len() {
        let name = archive.by_index(i)?.name().to_string();
        let rel = normalized_zip_relpath(&name)?;
        if rel.is_empty() {
            continue;
        }

        // Reject symlinks (best-effort via unix mode).
        let f = archive.by_index(i)?;
        if let Some(mode) = f.unix_mode() {
            let file_type = mode & 0o170000;
            if file_type == 0o120000 {
                return Err(MaterializeError::UnsafeZipEntry(name));
            }
        }

        ordered.push((rel, i));
    }
    ordered.sort_by(|a, b| a.0.cmp(&b.0).then_with(|| a.1.cmp(&b.1)));

    for (rel, idx) in ordered {
        let mut f = archive.by_index(idx)?;
        let out_path = dest_dir.join(&rel);

        if f.name().ends_with('/') {
            std::fs::create_dir_all(&out_path)?;
            continue;
        }
        if let Some(parent) = out_path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let mut out = std::fs::File::create(&out_path)?;
        // Stream copy to avoid buffering the full file in memory.
        std::io::copy(&mut f, &mut out)?;
    }

    Ok(dest_dir.to_path_buf())
}

#[derive(Debug, Clone)]
pub struct GitMaterializationResult {
    pub repo_root: PathBuf,
    pub head_commit: Option<String>,
}

pub fn materialize_git(
    git_url: &str,
    ref_name: Option<&str>,
    dest_dir: &Path,
) -> Result<GitMaterializationResult, MaterializeError> {
    // Deterministic: ensure the output directory is empty.
    if dest_dir.exists() {
        rmtree_retry(dest_dir, 30, Duration::from_millis(100))?;
    }
    if let Some(parent) = dest_dir.parent() {
        std::fs::create_dir_all(parent)?;
    }

    let repo = Repository::clone(git_url, dest_dir)?;

    if let Some(r) = ref_name {
        let obj = repo.revparse_single(r)?;
        repo.checkout_tree(&obj, Some(CheckoutBuilder::new().force()))?;
        repo.set_head_detached(obj.id())?;
    }

    let head_commit = repo
        .head()
        .ok()
        .and_then(|h| h.target())
        .map(|oid| oid.to_string());

    Ok(GitMaterializationResult {
        repo_root: dest_dir.to_path_buf(),
        head_commit,
    })
}
